File: java/0978_maxTurbulenceSize.java
class Solution {
    public int maxTurbulenceSize(int[] arr) {
        if(arr.length<2)
            return 1;
        else if(arr.length==2 && arr[0]!=arr[1])
            return 2;
        int right=1,left=0,res=1;
        while(right<arr.length-1)
        {
            if((arr[right]<arr[right+1]&&arr[right]<arr[right-1])||(arr[right]>arr[right+1]&&arr[right]>arr[right-1]))
                right++;
            else
                left = right++;
            if(!(left==right-1 && arr[left]==arr[right]))
                res = Math.max(res,right-left+1);
        }
        return res;
    }
}

File: cpp/0978_maxTurbulenceSize.cpp
// sliding window:右指针正常右移，左指针主动右移（需判断条件）; 记res为最大长度
// 右指针：当符合湍流条件时，右移
// 左指针：当不符合湍流条件时，下一个可能出现的子数组一定在右指针位置之后，因此直接移动到和右指针相同的位置，右指针正常右移
// 特殊情况：经过上述移动后，左右指针相邻并且值一样，此时不需要更新res
class Solution {
public:
    int maxTurbulenceSize(vector<int>& arr) {
        if(arr.size()<2)
            return arr.size();
        else if(arr.size()==2)
            return arr[0]==arr[1]?1:2;
        int res=1,right=1,left=0;
        while(right<arr.size()-1)
        {
            if((arr[right]>arr[right-1] && arr[right]>arr[right+1])||(arr[right]<arr[right-1] && arr[right]<arr[right+1]))
                right++;
            else
                left = right++;
            if(!(right==left+1 && arr[right]==arr[left]))
                res = max(res,right-left+1);
        }
        return res;
    }
};
